linux_os/guide/system/network/network-firewalld/ruleset_modifications/ensure_firewall_rules_for_open_ports/oval/shared.xml

```diff
@@ -1,76 +1,183 @@
+    <filter action="exclude">state_ipv6_loopback_listening_inet_ports</filter>
+    <filter action="exclude">state_inet_foreign_port_connected</filter>
+  </linux:inetlisteningservers_object>
 
+  <local_variable id="var_obj_listening_inet_{{{ proto }}}_ports" version="1"
+    comment="Variable with all firewalled ports" datatype="int">
+    <object_component object_ref="obj_listening_inet_{{{ proto }}}_ports" item_field="local_port"/>
+  </local_variable>
 
-  <linux:inetlisteningservers_state id="state_ipv4_loopback_listening_inet_ports" version="1" comment="Checks local address is not ipv4 loopback." xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#linux">
-    <linux:protocol operation="pattern match">^.*$</linux:protocol>
-    <linux:local_address operation="equals">127.0.0.1</linux:local_address>
-    <linux:local_port datatype="int" operation="greater than or equal">0</linux:local_port>
-  </linux:inetlisteningservers_state>
+  <ind:variable_object id="object_var_obj_listening_inet_{{{ proto }}}_ports" version="1">
+    <ind:var_ref>var_obj_listening_inet_{{{ proto }}}_ports</ind:var_ref>
+  </ind:variable_object>
 
-  <linux:inetlisteningservers_state id="state_ipv6_loopback_listening_inet_ports" version="1" comment="Checks local address is not ipv6 loopback." xmlns="http://oval.mitre.org/XMLSchema/oval-definitions-5#linux">
-    <linux:protocol operation="pattern match">^.*$</linux:protocol>
-    <linux:local_address operation="equals">::1</linux:local_address>
-    <linux:local_port datatype="int" operation="greater than or equal">0</linux:local_port>
-  </linux:inetlisteningservers_state>
+  <ind:textfilecontent54_object id="object_firewalled_service_{{{ proto }}}_port" version="1">
+    <ind:path>/usr/lib/firewalld/services/</ind:path>
+    <ind:filename operation="pattern match" var_ref="var_all_firewalled_services_desc_filename"
+      var_check="at least one"/>
+    <ind:pattern operation="pattern match">\s*(?:(?:protocol="{{{ proto }}}")|)\s*port="(\d+)"\s*(?:(?:protocol="{{{ proto }}}")|)\s*</ind:pattern>
+    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
 
-  <ind:textfilecontent54_object id="object_firewalled_service" version="1">
+  <local_variable id="var_firewalled_service_{{{ proto }}}_ports" version="1"
+    comment="Firewalled ports according to firewalld configuration per service" datatype="int">
+    <object_component object_ref="object_firewalled_service_{{{ proto }}}_port" item_field="subexpression"/>
+  </local_variable>
+
+  <ind:variable_object id="object_var_firewalled_service_{{{ proto }}}_ports" version="1">
+    <ind:var_ref>var_firewalled_service_{{{ proto }}}_ports</ind:var_ref>
+  </ind:variable_object>
+
+  <!-- States objects and variables describing the Firewalled ports stated with port numbers -->
+  <ind:textfilecontent54_object id="object_firewalled_direct_{{{ proto }}}_ports" version="1">
```

Comment:
It would be great to have a tests scenario also using direct ports instead of referencing a service. It would test this case here.